### ISSUE TITLE
Support passing separate validation data set to pytorch-lightning based models

### ISSUE THREAD

**Issue by MischaPanch** [jambit/sensAI#55]
With the higher level vector model interfaces, validation in training of neural networks can currently only be performed with a train-test split. On the other hand, the lower level NNOptimizer itself supports passing a separate validation set.  Pytorch-lightning's `Trainer` class also has similar capabilities which are currently disabled in sensai because of vector model's fit interface. I ran into this myself because I want to fit pytorch-lightning models to a data frame where the validation set cannot be created by splitting (due to data leakage).

Note that it is not sufficient to use evaluators for performing validation on separate sets since this way the validation set cannot be used for early stopping.

I propose to address this by relaxing VectorModel's interface and allowing to pass a validation set in `fit` and `fit_classifier`.

Something like `fit(X, Y, X_validation=None, Y_validation=None)`

@opcode81 @schroedk 
If you think this is reasonable, I would prepare a PR in the next days.

**Comment by opcode81** [comment]
I don't think it is reasonable, because validation sets are not usable in the general case and therefore have no place in VectorModel. 
If you require a custom split, implement a TorchDataSetProvider that performs it according to your requirements (and override the method that creates the provider in your model implementation). Alternatively, use the low-level interface like you mentioned.

**Comment by MischaPanch** [comment]
I am not using the torch models, so unfortunately this is not an option. I will probably slightly extend sensAI's pytorch-lightning interfaces in a PR then, atm. they are rather tightly bound to vector model. You are right, for general models validation sets are not needed at fit time, so it's not good to put that to vector model. I changed the issue title accordingly

I'll try to find time to make a notebook documenting the lower level torch interfaces, including the provider-overriding strategy you mentioned, which seems to be the only way for doing custom splits and still using the perks of vector model

**Comment by opcode81** [comment]
I guess the question then is: Why aren't you using the torch models? They are extremely flexible.

**Comment by MischaPanch** [comment]
Well, partly because I'm training a CNN, and that's not too much supported in the torch models, if I remember correctly. But the main reason is simply because I wanted to try out the pytorch lightning modules within sensai in a project. I expect to encounter some model serialization issues on the way which I wanted to resolve at some point any way. This particular issue will be easy to resolve, I think. I will just allow to pass a custom data splitter to the instantiation of `PLWrappedModel` and all classes that use it.

Btw, making the data splitting as part of the model implementation by overriding the corresponding does not feel all too good. The data handling should not be delegated to the model, right? (Passing in a custom splitter that would not work on all data frames also violates this, but seems a bit more benign)

Using the low level torch interfaces for model fitting significantly decreases the features I can use, in particular the evaluation module and the whole feature normalization and extraction utils. Or am I missing something?

**Comment by opcode81** [comment]
There is no problem with CNNs. Why would there be?

How much data handling should be delegated to your model depends on the generality of your model and the degree to which it is independent of input data. If your model expects certain/specific input data,

**Comment by opcode81** [comment]
I bloody despise mobile devices. Still in Italy 🇮🇹

**Comment by MischaPanch** [comment]
My model is completely generic (just a CNN) but my data frame would have data leakage to the validation set if I were to perform a random split. I can very easily create a validation data frame by doing some extremely specific filtering that has to do with how I generate the data and that might change at any moment (I could also simply load the validation data frame separately). Putting this logic directly in the model implementation does not feel right to me. But it might well be that this use case is too specific to be fully supported everywhere in sensAI :)

**Comment by opcode81** [comment]
It seems to me that you have a generic model and a specific use case for that model where you also have specific data. The latter can be represented in its own subclass that is no longer generic.

**Comment by opcode81** [comment]
By the way, in the sensai.torch way of doing things, the generic low-level model would be a TorchModel or even torch.nn.module and the application-specific stuff would be a Torch*VectorModel handling all the data-specific stuff.

**Comment by opcode81** [comment]
Note that `Torch*VectorModel` classes have recently been extended to extensively support dependency injection in order to modify behaviour, including the provision of a mechanism for splitting the training set into a validation set and the the actual test set (via a `TorchDataSetProviderFactory`); overrides are no longer required or encouraged. 
Moreover, the default factory now supports parametrisation of the splitting mechanism by injecting a `DataFrameSplitter`, a component which can be used for all splits (i.e. not only the train-validation split but also the train-test split as applied within evaluators).

So can we close this issue?

**Comment by MischaPanch** [comment]
Cool, glad to hear that the splitter is injectible now! Let me implement the same mechanism for the pytorch-lightning modules, I will make a PR tomorrow. Then we can close this

**Comment by opcode81** [comment]
Closing because the pytorch lightning based stuff is essentially abandoned.
Going forward, I think we should think about how we could extend the current torch-based models to support lightning-based models (if there is a need for it). Specifically, this would involve supporting LightningModule and Trainer as alternatives to TorchModel and NNOptimiser respectively.
